gobject-introspection: add patch to fix in-build scanner execution

michaelld opened the pull request:
Ref: https://trac.macports.org/ticket/58574

Hopefully this PR addresses that ticket's issue, finally. I note "Ref" since I want it referenced, but not necessarily closed. Hopefully closed but only after extensive testing!

#### Description

Various tweaks to the GI-O scanner Python code to provide for in-build execution. Getting this into a PR for everyone to test / verify. The code changes read well to me, and should be macOS-version agnostic.

Note that since we're using an older version of GO-I (pre-meson @ 1.60.2 ; current is 1.66.1), so any fixes here may or not be relevant to current GO-I. If/when we decide to upgrade GO-I, it will involve retesting and possible fixing up the patches we come up with to fix this issue for us ... which makes upgrading that much more challenging!

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
printf "%s\n" "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)" "$(xcodebuild -version|awk 'NR==1{x=$0}END{print x" "$NF}')"|tee /dev/tty|pbcopy
-->
macOS 11.0 20A5384c
Xcode 12.2 12B5025f

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

kencu commented:
Does ```LIBRARY_PATH``` do anything on darwin? I'm not used to manipulating that envvar.

I would think we should not use DYLD_FALLBACK_LIBRARY_PATH at all...

I would think we should use DYLD_LIBRARY_PATH, but only add in-tree paths to it (somehow or other).

michaelld commented:
> Does `LIBRARY_PATH` do anything on darwin? I'm not used to manipulating that envvar.

Yes; yes it does! We use it in MacPorts in plenty of places. Looks at qt4-mac!
 
> I would think we should not use DYLD_FALLBACK_LIBRARY_PATH at all...

Yeah I'm thinking that too. I'm pretty sure the `LIBRARY_PATH` is what's going the trick in that section of code. I'll test and confirm.
 
> I would think we should use DYLD_LIBRARY_PATH, but only add in-tree paths to it (somehow or other).

That's exactly what I'm doing: DYLD_LIBRARY_PATH for in-build paths ; LIBRARY_PATH for all other paths. That combination seems to do the trick here!

michaelld commented:
> Yeah I'm thinking that too. I'm pretty sure the `LIBRARY_PATH` is what's going the trick in that section of code. I'll test and confirm.

Confirmed. It's not clear that it hurts having the FALLBACK variable set, but it probably doesn't help here since the setting is primarily being used for compiler linking purposes, not binary execution purposes.

ryandesign commented on gnome/gobject-introspection/files/patch-fix-scanner-in-build-execution.diff:
```diff
@@ -0,0 +1,61 @@
++            runtime_path_envvar = ['LIB', 'PATH']
++        else:
++            if sys.platform == "darwin":
++                runtime_path_envvar = ['LIBRARY_PATH','DYLD_FALLBACK_LIBRARY_PATH']
```
I don't have experience with non-Mac operating systems, but on macOS, at runtime, the dynamic library loader will look for libraries first in the locations specified in `DYLD_LIBRARY_PATH`, then in the locations baked into the executable or library, then in locations specified in `DYLD_FALLBACK_LIBRARY_PATH`. I agree with Ken that we don't have any need to look in `DYLD_FALLBACK_LIBRARY_PATH` at any time.

`LIBRARY_PATH` is not used at runtime. Specifying a path in `LIBRARY_PATH` is similar to specifying it with an `-L` flag: it is used by the linker at link time to find libraries. I'm not sure what dumper.py does. Does it invoke the linker? If so, then maybe it does need to set environment variables that will be used by the linker, and it's just a case of a misnamed variable (`runtime_path_envvar`).

michaelld replied:
This specific chunk of `dumper.py` is setting up the link command and environment. Hence using `LIBRARY_PATH` is the way to go for Darwin / macOS and why I split this setting off from other non-Windows settings. Any `DYLD` settings here will not impact the link.

That said, settings here are carried forth throughout the rest of the program. Hence minimizing the settings here is best. I will remove the `FALLBACK` portion, as I've tested this both with and without it and it works fine both ways, and hence that portion is not required.

michaelld replied:
FYI `poppler`s CMake script adds on the paths for the local build and source to find both headers and libraries. gO-I correctly sets those paths for the build and link commands. What I'm fixing here is the use of "system" library search path, in this case defined as paths that are not in the build paths somewhere.

The overall code flow here is to glean library linkage information using `pkg-config` and add it to the general link command. This is fine -except- for the inclusion of any `-L` "system" paths, which are better off being in `LIBRARY_PATH` instead of on the link command. This be because prior `-l` libraries are in-build -- with the proper `-L` to find them -- and those libraries might have an installed version already in the "system" paths. Hence moving the "system" library paths from `-L` to `LIBRARY_PATH` allows the linker find the in-build version instead of the already-installed "system" version ... which is what this `poppler` build issue is all about in the first place!

ryandesign replied:
Reading dumper.py now, I see that this code you're changing is inside a function called `_link` which does call the linker.

Before it was asking pkg-config for all flags, including `-L` flags, and adding those to project-specific flags, including `-L` flags, and then calling the linker with all those args. Now you've changed it so that the pkg-config `-L` flags get changed into paths in `LIBRARY_PATH` instead, which should work the same except for the very old versions of clang that don't support `LIBRARY_PATH` that we don't care about anymore.

What I don't see yet is where those args are carried forward to the rest of the program. Where's that?

michaelld replied:
> What I don't see yet is where those args are carried forward to the rest of the program. Where's that?

Not sure what you mean here. I'll take a guess:

Once the `os.environ` is set, those variables are carried forth inside Python until they are either changed / unset or Python exits. Unlike local variables inside Python (or C/C++ or most languages), these variables are set for the overall runtime environment and hence if we don't want them carry forth we'll need to save the old environment, modify it, and restore it at the end. Not difficult to do, but I also don't see a great need for doing so.

ryandesign replied:
>> What I don't see yet is where those args are carried forward to the rest of the program
>
> Not sure what you mean here

Well you said: 

> settings here are carried forth throughout the rest of the program

But I should probably stop asking questions and just test your changes first.

michaelld replied:
> > settings here are carried forth throughout the rest of the program

Yes ... as noted in my reply and augmenting here: Unlike local variables inside Python (or C/C++ or most languages), these variables are set for the overall Python runtime environment. Hence, setting them here will carry them forth beyond just this specific function into the code that called it, and the code that called that code, etc -- until either the runtime environment is changed or Python returns from execution.

One option we have is to modify the code to save the runtime environment, change it, do the link, then restore the runtime environment back to the state it was in at the start of the function. Doing this won't be too difficult, but I also don't see doing it as necessary.

> But I should probably stop asking questions and just test your changes first.

Testing the changes is definitely in order here. The code isn't the most obvious. I added a ton of debug `print` statements to see what was going on throughout the code execution, to get to the point of these changes. I'm confident that these changes do the trick for MacPorts' needs. But any testing I did cannot be exhaustive & hence I might have missed something. Thus, yes, I appreciate others testing / verifying that these changes work for both `poppler` as well as don't impact other ports that use gO-I .

michaelld commented:
I just pushed an update to remove the `DYLD_FALLBACK_LIBRARY_PATH` as already noted.

neverpanic commented on gnome/gobject-introspection/files/patch-fix-scanner-in-build-execution.diff:
```diff
@@ -0,0 +1,61 @@
+         args.extend(sources)
+ 
+-        pkg_config_libs = pkgconfig.libs(self._packages,
++        pkg_config_libs, pkg_config_libs_only_L = pkgconfig.libs(self._packages,
```
Before submitting this upstream, you'll likely want to fix the indentation, either by moving the `msvc_syntax` param in the next line to the same column as the first argument of the call to `pkgconfig.libs()`, or, since the line is getting long anyway, by moving `self._packages` indented to the next line.

michaelld replied:
Here's my proposal on submitting upstream:

Given how far our GO-I is behind current release / master, I'm not going to take the time to work out -if- this is even an issue with the current release / GIT master.

-When- we update GO-I to a meson-based release -- hopefully the current GO-I release -- then I'll work through this issue again determine whether any changes are needed, and if so then update the patches here in MP -- as well as submit them upstream.

michaelld replied:
Agreed on the reformatting. Given that this patch is a one-off & not to be submitted upstream (IMHO at least), I'm OK with the formatting.

neverpanic commented on gnome/gobject-introspection/files/patch-fix-scanner-in-build-execution.diff:
```diff
@@ -0,0 +1,61 @@
+                                          msvc_syntax=self._compiler.check_is_msvc())
++        this_L = []
++        for _pc_libs_L in pkg_config_libs_only_L:
++            this_L.append(_pc_libs_L[len('-L'):])
```
```python
this_L = [lib[len('-L':)] for lib in pkg_config_libs_only_L]
```

does the same thing, is shorter, and arguably more idiomatic python.

michaelld replied:
oh nice! I'm not a Python guru ... I know the basics & can sometimes come up with these shortcuts ... sometimes not. Thx!

michaelld replied:
Minor fix: `len('-L':)` -> `len('-L'):`

neverpanic reviewed (commented):
Thanks for working on this – I always thought it's probably not that hard to fix this, but did not get around to doing it. Some nitpicks on the Python.

One question though: From what I see, this will now set `DYLD_LIBRARY_PATH` to the paths returned by `pkg-config --libs-only-L`, but won't that get us into trouble? For example, let's assume we're trying to run gobject-introspection on a library that links against `ImageIO.framework` from `/System/Library/Frameworks`. That in turn links against its own copy of libjpeg, but setting `DYLD_LIBRARY_PATH` to `/opt/local/lib` as often returned by pkg-config will make it load `/opt/local/lib/libjpeg.dylib` instead and fail?

In the discussion I see mentions of only adding in-build paths to `DYLD_LIBRARY_PATH`, which would likely avoid the problem, but I don't see where that's implemented. What am I missing?

michaelld commented:
> Edit: Nevermind, I see it now. You're not setting DYLD_LIBRARY_PATH in the cases where this would be a problem.

Correct! `DYLD_LIBRARY_PATH` is purely for the in-build libraries, and is just for execution purposes; hence no "system" libraries. `LIBRARY_PATH` is used for "system" libraries, and of course is just for linking purposes.

michaelld commented:
I'm going to fixup the Python per request, rebase, squash, re-test, and force push the updated changes.

michaelld commented:
I just updated the PR:

* rebased to current master ;

* squashed all commits ;

* tightened up the Python as noted with my code fix due to the typo in the suggestion ;

* reduced the `_link` variable setting for Unix-y system to just `LIBRARY_PATH`

Please keep testing & commenting & discussing! I don't expect us to update GO-I any time soon, so we can afford some time to do thorough testing here without worrying about the changes going stale! That said, if we go a week without further comment then I'm inclined to merge this PR and see if it works in the wide wild world :)

neverpanic reviewed (approved):
This lg2m!

kencu commented:
Sorry my review is tardy -- I don't really speak much python, to be honest, and it's a bit Greek to me. Just so I know where to start, I assume there are two different lists being kept, one to go onto LIBRARY_PATH for the compiler, and one to go onto DYLD_LIBRARY_PATH for runtime. I can't quite see which list is which, though -- and I can't quite make out how what you've done here will make the in-tree paths go onto DYLD_LIBRARY_PATH.

If you have a moment to explain which variable is which, that would be appreciated. If not, I'm fine with everyone else's thumbs-up and we'll see how it works out!

Thanks for taking the time, Michael -- people were starting to get a bit rabid about this, as poppler updates every few days it seems...

michaelld commented:
@kencu a "quick" note about those 2 variables and my understanding how they are used, FYI and per your request ... others please chime in here if I'm missing something or mistaken somewhere ...

1) `LIBRARY_PATH` is used by the linker. When provided a library to find the path to, the linker first searches internally-known paths (unless told not to; there's a flag for this), then those supplied with `-L`, and finally any in `LIBRARY_PATH`.

--> The issue currently is that the `-L` entries contain both the build directories and the "system" paths, and thus which is selected depends on the linker and how it determines the search order -- which for whatever reason for `poppler` means the "system" ones first. We don't want this! We want to force the linker to search the "system" paths only after all other paths have been searched.

--> By putting the pkg-config `-L` paths -- "system" paths -- into the `LIBRARY_PATH` -- and removing them from the `-L` linker arguments -- we allow the linker to search those paths, but only after exhausting all other path options. Since there will be `-L` path entries for just-built libraries (in the build directory somewhere), libraries that exist in both the build directory and `LIBRARY_PATH` will be found in the build only; never in the "system" paths. Guaranteed.

2) `DYLD_LIBRARY_PATH` is used by the DYLD runtime. When you try to run an executable, there are libraries linked to it that have to be found. These libraries are shown via `otool -L`, and contain the whole path to each of them or an `@rpath` -- let's ignore this latter for the moment. DYLD will search any valid paths in the `DYLD_LIBRARY_PATH` -first-, then search the paths embedded in the executable, then search any valid paths in the `DYLD_FALLBACK_LIBRARY_PATH`.

--> The just-built libraries -- which will be installed into the "system" paths -- internally contain the "system" paths to the libraries already. Thus, linking against them copies the "system" paths to the executable. If `DYLD_LIBRARY_PATH` is not set, then if the libraries contained in the executable are found in the "system" paths then they will be used.

--> To make sure the temporary GO-I executable uses the just-built libraries in the build directories, we must set `DYLD_LIBRARY_PATH` to contain the correct build paths for those libraries, otherwise those found in the executable itself will be used -- which in this case are "system" paths.

michaelld commented:
After some reflection, I should also note that:

A) There are other variables and settings that impact both the linker and DYLD (and Linux LD) searches. But, to a first order what I wrote is what makes the most difference for this PR.

B) I think the linker -- whether called from GCC or Clang, whether Darwin or GNU/Linux -- works roughly the same in terms of the search paths & variable and flags influencing the search ordering.

C) As a comparison to Darwin's DYLD, the Linux LD uses a cache to store the information about known libraries -- after any install, one must update the cache to find the newly installed library/ies ... a bit of a pain, but then when said library is referenced it will always be found -- not the case with Darwin's DYLD! When an executable is run on Linux, my memory is that the `LD_LIBRARY_PATH` is searched first, then the cache is used ... I don't recall if there is a `FALLBACK` or not. as noted in (A), I'm sure there are other variables that influence LD.

D) I believe for DYLD there is an environment variable one can set to provide debug information as DYLD does it's searching. It's been quite a while since I read though the DYLD manual / how-to guide, but my memory is that there are plenty of DYLD environment variables that show information and/or influence how DYLD does it's business! I'd guess the same is true for Linux LD; I'm not as experienced in it as I am with DYLD ... but I'm getting there as my for-pay work takes me further into Linux systems!

michaelld commented:
24 hour notice to me merging this if nobody else does and/or if there are not justifiable objections to doing so.
